### DISCUSSION TITLE
Warming up a worker's pytorch/setfit model

### DISCUSSION THREAD

**Discussion by lsmith77** [fastapi/fastapi#9333] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# in the global namespace
from setfit import SetFitModel

setfit_model = SetFitModel.from_pretrained(pretrained_lang_model)
# pre-warm the model
setfit_model(["Hello guys"])

def check_context_valid(sentence):
    context = setfit_model([sentence])
    return context.item() == 1
```

### Description

when running with gunicorn "--preload" the pre-warm has no effect and the requests timeout and gunicorn restarts the worker with singal 6.

if I disable "--preload" it works "fine" but I then run into issues as I need so much memory as each worker will load all the models into their own memory.

### Operating System

Linux

### FastAPI Version

0.93.0

### Python Version

3.11.1

### Additional Context

Note I am running on CPU's

**Comment by jgould22** [comment]
How long does loading and pre warming the model take ?

Gunicorn will only wait 30s for a worker to start, if it is taking longer than that to load and warm your model you will need to give it a longer [timeout](https://docs.gunicorn.org/en/stable/settings.html#timeout).

**Reply by lsmith77** [reply]
I have increased the timeout already. But the issue is not the initial load per se.

If I have "--preload" enabled, I can load the model (again I do this in the global space). I can access any endpoint that doesn't access the model. But as soon as I access the model, I get the worker timeout/signal 6.

Which mean the issue is not about getting the worker up and running. The issue is that the first request that access the model takes too long.

**Reply by jgould22** [reply]
So presumably when gunicorn forks the process for the worker something is not quite right. 

I do not know too much about how the model bits work but does your warming persist if you were to fork the model yourself ? That would probably help you eliminate the anything happening in your model lib from being the problem.

Are you able to use one of the [gunicorn hooks](https://docs.gunicorn.org/en/latest/settings.html#post-fork) to run your warming code ?

**Comment by jvllmr** [comment, marked as answer]
Have you tried using a lifespan handler (https://fastapi.tiangolo.com/advanced/events/?h=event#lifespan-function)?

The example in the docs actually loads a machine learning model.
FastAPI just recently changed the API for this. If you're running on an older version, this is the deprecated usage:
https://fastapi.tiangolo.com/advanced/events/?h=event#startup-event

**Reply by lsmith77** [reply]
yes I have tried that as well. I think it fundamentally isn't different for this use case to doing the warming in the global space as shown in the code in my original post.
